spring-cloud-task-core/src/main/java/org/springframework/cloud/task/listener/annotation/TaskListenerExecutor.java

```diff
@@ -0,0 +1,246 @@
+/*
+ * Copyright 2016 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.task.listener.annotation;
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.aop.framework.autoproxy.AutoProxyUtils;
+import org.springframework.aop.scope.ScopedObject;
+import org.springframework.aop.scope.ScopedProxyUtils;
+import org.springframework.beans.factory.BeanInitializationException;
+import org.springframework.beans.factory.config.ConfigurableListableBeanFactory;
+import org.springframework.cloud.task.listener.TaskExecutionListener;
+import org.springframework.cloud.task.repository.TaskExecution;
+import org.springframework.context.ConfigurableApplicationContext;
+import org.springframework.core.MethodIntrospector;
+import org.springframework.core.annotation.AnnotationUtils;
+
+/**
+ * Identifies all beans that contain a TaskExecutionListener annotation and stores the
+ * associated method so that it can be called by the {@link TaskExecutionListener} at the
+ * appropriate time.
+ *
+ * @author Glenn Renfro
+ */
+public class TaskListenerExecutor implements TaskExecutionListener{
+
+	private ConfigurableApplicationContext context;
+
+	private final static Logger logger = LoggerFactory.getLogger(TaskListenerExecutor.class);
+
+
+	private final Set<Class<?>> nonAnnotatedClasses =
+			Collections.newSetFromMap(new ConcurrentHashMap<Class<?>, Boolean>());
+
+	private Map<Method, BeforeTask> beforeTaskMethods;
+	private Map<Method, Object> beforeTaskInstances;
+
+	private Map<Method, AfterTask> afterTaskMethods;
+	private Map<Method, Object> afterTaskInstances;
+
+	private Map<Method, FailedTask> failedTaskMethods;
+	private Map<Method, Object> failedTaskInstances;
```

Comment:
I may not be following here, but why do you need to store a `Map<Method, ANNOTATION>` and `Map<Method, INSTANCE>`?  Since you have a map for each annotation type, why isn't the latter sufficient?